### PR TITLE
Stabilise SIP-47

### LINKED ISSUE #20769: Stabilise SIP 47 - Clause Interleaving
referencing comment https://contributors.scala-lang.org/t/when-will-sip47-clauseinterleaving-graduate-beyond-experimental/6686/3 - it is now considered shipped

## Actions

remove `scala.language.experimental.clauseInterleaving` requirement from `3.6`

### PR DESCRIPTION
**Pull request by hamzaremmal**
Closes #20769 

Initial implementation in #14019

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala:3881]: shouldn't this still be disabled when source version is below 3.6?

Comment (hamzaremmal) [reply compiler/src/dotty/tools/dotc/parsing/Parsers.scala:3881]: It should yes

**Comment by hamzaremmal** [comment]
This based on top of #20895

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala
Comment (hamzaremmal) [reply compiler/src/dotty/tools/dotc/parsing/Parsers.scala:3881]: @bishabosha I've also added a test to check the correct behavior with `-source 3.5`

Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/parsing/Parsers.scala:3881]: nice

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala
Comment (odersky) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala:3881]: This looks out of line with other code that's version dependent, Why the `using in.languageImportContext`?
That's done nowhere else in the parser. Why not just the original test?

Comment (odersky) [reply compiler/src/dotty/tools/dotc/parsing/Parsers.scala:3881]: There's a doc comment further up that needs to be updated. 
```

     * if clauseInterleaving is enabled:
     *  DefSig  ::=  id [DefParamClauses] [DefImplicitClause]
     * 
  ```
As far as Parser is concerned, clauseInterleaving _is_ now enabled, so the grammar should reflect this.

**Review by odersky — CHANGES_REQUESTED** [review]
Otherwise LGTM

**Review by odersky — APPROVED** [review]

**Comment by hamzaremmal** [comment]
I will wait for #20895 to be merged before merging this PR

File: compiler/src/dotty/tools/dotc/config/Feature.scala
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/config/Feature.scala:125]: this shouldn't be mixing source version with language import checks, they should be kept separate,
e.g. in the parser do `sourceVersion.isAtLeast(`3.6`) || in.featureEnabled(Feature.clauseInterleaving)`

Comment (hamzaremmal) [reply compiler/src/dotty/tools/dotc/config/Feature.scala:125]: not sure to see why we should do that ?

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala:3878]: like I said earlier, language imports and source versions should be kept separate 
```suggestion
          if sourceVersion.isAtLeast(`3.6`) || in.featureEnabled(Feature.clauseInterleaving) then
```

File: compiler/src/dotty/tools/dotc/config/Feature.scala
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/config/Feature.scala:125]:
```suggestion
    enabled(clauseInterleaving)
```

Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/config/Feature.scala:125]: ok, I guess its fine then, curious that in parser they just have `Feature.fewerBracesEnabled` without explicit using

Comment (hamzaremmal) [reply compiler/src/dotty/tools/dotc/config/Feature.scala:125]: Yea, I was currious too. I guess it might be a no-op as it is, therefore a bug ?

**Comment by Gedochao** [comment]
This has been decided to be included in the 3.6.0 release.

**Comment by SethTisue** [comment]
should https://github.com/scala/scala3/discussions/21195 be a blocker for this?

**Comment by sjrd** [comment]
I don't think so. It only affects new code combined with Java code. It cannot break anyone immediately. We should fix it, though, for sure, but not a blocker IMO.

**Comment by SethTisue** [comment]
okay, it was open as a discussion but I've made it into an issue, #21346

**Pull request merged**
